### PR TITLE
Add Adhoc Task for Each Course Module

### PR DESCRIPTION
**Pull request by rajandangi**
This update introduces an adhoc task for each course module.

- Added a new file `classes/task/reengagement_adhoc_task.php` to handle the adhoc task logic.
- Moved portion of code form `lib.php` to integrate the adhoc task functionality.

### CODE REVIEW

**Comment by rajandangi** [comment]
The CI failed due to deprecated 404 Behat tests, which aren't related to this change. However, I've addressed the 404 test issue in [this branch](https://github.com/catalyst/moodle-mod_reengagement/commits/fix/behat_404_test/). If you prefer, we can create a separate branch specifically for the 404 fix, including the updated Behat test."

File: classes/task/reengagement_adhoc_task.php [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+        $data = $this->get_custom_data();
+        if (empty($data)) {
+            return;
+        }
+
+        // Get a list of users who are eligible to start this module.
+        $startusers = reengagement_get_startusers($data);
+
+        // Prepare the objects for db iteratin.
```
Comment (petersistrom) [inline classes/task/reengagement_adhoc_task.php]: Just a tiny typo in the comment should be `iteration`

**Comment by petersistrom** [comment]
Hi @danmarsden 
I've reviewed and tested @rajandangi's PR and it looks good to me, no new code was added it's just splitting the scheduled task into adhoc tasks for each course module. Just checking in with you to make sure that in principle it is something you are happy to have merged in?

**Comment by danmarsden** [comment]
no objections to the concept - if someone has carefully peer reviewed the patch and made sure it does pretty much the same thing - of course this needs to land in the latest branch "first" before the 400_STABLE branch is merged in but happy for you to hit the merge button on your end if that's done.

**Comment by rajandangi** [comment]
Thanks, @danmarsden , for creating the new MOODLE_404_STABLE branch.

Since we now have a dedicated branch for 404, all the tests are passing on this PR. Please check and confirm.

**Comment by rajandangi** [comment]
> no objections to the concept - if someone has carefully peer reviewed the patch and made sure it does pretty much the same thing - of course this needs to land in the latest branch "first" before the 400_STABLE branch is merged in but happy for you to hit the merge button on your end if that's done.

I’ve created another PR #174  for the latest branch by cherry-picking the commit that splits the scheduled task into an adhoc_task for each course module. Please check and confirm.

**Pull request merged**